src/storage-client/src/storage_collections.rs

```diff
@@ -0,0 +1,2640 @@
+use timely::progress::{Antichain, ChangeBatch, Timestamp as TimelyTimestamp};
+use timely::PartialOrder;
+use tokio::sync::{mpsc, oneshot};
+use tracing::{debug, info, trace, warn};
+
+use crate::controller::{
+    CollectionDescription, DataSource, DataSourceOther, PersistEpoch, StorageMetadata, StorageTxn,
+};
+
+/// An abstraction for keeping track of storage collections and managing access
+/// to them.
+///
+/// Responsibilities:
+///
+/// - Keeps a critical persist handle for holding the since of collections
+///   where it need to be.
+///
+/// - Drives the since forward based on the upper of a collection and a
+///   [ReadPolicy].
+///
+/// - Hands out [ReadHold] that prevent a collection's since from advancing
+///   while it needs to be read at a specific time.
+#[async_trait(?Send)]
+pub trait StorageCollections: Debug {
+    type Timestamp: TimelyTimestamp;
+
+    /// On boot, seed metadata/state.
+    async fn initialize_state(
+        &mut self,
+        txn: &mut dyn StorageTxn<Self::Timestamp>,
+        init_ids: BTreeSet<GlobalId>,
+        drop_ids: BTreeSet<GlobalId>,
+    ) -> Result<(), StorageError<Self::Timestamp>>;
+
+    /// Resets the txns system to a set of invariants necessary for correctness.
+    ///
+    /// Must be called on boot before create_collections or the various appends.
+    /// This is true _regardless_ of whether the persist-txn feature is on or
+    /// not. See the big comment in the impl of the method for details. Ideally,
+    /// this would have just been folded into `Controller::new`, but it needs
+    /// the timestamp and there are boot dependency issues.
+    ///
+    /// TODO: This can be removed once we've flipped to the new txns system for
+    /// good and there is no possibility of the old code running concurrently
+    /// with the new code.
+    async fn init_txns(
+        &mut self,
+        init_ts: Self::Timestamp,
+    ) -> Result<(), StorageError<Self::Timestamp>>;
+
+    /// Marks the end of any initialization commands.
+    ///
+    /// The implementor may wait for this method to be called before
+    /// implementing prior commands, and so it is important for a user to invoke
+    /// this method as soon as it is comfortable. This method can be invoked
+    /// immediately, at the potential expense of performance.
+    fn initialization_complete(&mut self);
+
+    /// Update storage configuration with new parameters.
+    fn update_parameters(&mut self, config_params: StorageParameters);
+
+    /// Returns the [CollectionMetadata] of the collection identified by `id`.
+    fn collection_metadata(
+        &self,
+        id: GlobalId,
+    ) -> Result<CollectionMetadata, StorageError<Self::Timestamp>>;
+
+    /// Returns the since/upper frontiers of the identified collection.
+    fn collection_frontiers(
+        &self,
+        id: GlobalId,
+    ) -> Result<
+        (Antichain<Self::Timestamp>, Antichain<Self::Timestamp>),
+        StorageError<Self::Timestamp>,
+    >;
```

Comment:
this should be a provided method built on top of `collections_frontiers`

Reply:
will do!